feat(tabs)!: add delete tabs

savutsang opened the pull request:
[DS-764](https://equisoft.atlassian.net/browse/DS-764)

**Changes**
- Added a feature to allow to add and remove tabs.
- New modern design for the Tabs.
- Rework on the design tokens to reflect the new design.

**Others**
- Improved the test utility findByTestId/getByTestId to be able to target more precisely the wanted node (ex: mount() wrapper might create 2-3 nodes with the same data-testid).

[DS-764]: https://equisoft.atlassian.net/browse/DS-764?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BREAKING CHANGE:
- Removed "contained" prop. The new design don't allow this customization anymore.

LarryMatte reviewed (changes requested):
Quelques commentaires ici et là.

Sinon quelques modifications que j'apporterais au comportement clavier:

- J'enlèverais la possibilité de focus (touche Tab du clavier) sur les "x" des tabs qui ne sont pas actives.
- Dans la story "normal" quand j'utilise la touche Tab du clavier, le focus passe du **Tab** au **TabPanel** et après il devrait tomber sur le bouton de sorting dans le tableau mais il ne s'y rend pas. Comme si la navigation avec la touche Tab skippait ce qui se trouve dans le **TabPanel**.

Prochain point est mon erreur... my bad
lorsqu'un utilisateur ajoute plusieurs Tabs, les Tabs s'enlignent les unes à la suites des autres et elles deviennent toutes compressées. Dans [Figma](https://www.figma.com/file/EIJItUlyeIymb2dfgeoQJp/Design-System%2C-Desktop?type=design&node-id=8586-1421&mode=dev), il y a 2 boutons placés au début et à la fin du Tablist lorsque les Tabs dépassent l'espace prévu par le Tablist.
Il faudrait ajouter la features dans la PR. Toutes mes excuses. Il y a un exemple similaire dans le Design system de [Carbon](https://react.carbondesignsystem.com/?path=/story/components-tabs--contained-full-width&globals=theme:white)

maboilard reviewed (changes requested):
Mes commentaires sont pas mal similares à ceux de Larry. 

Il manquerait les boutons en cas d'overlflow (voir [ici](https://www.figma.com/file/EIJItUlyeIymb2dfgeoQJp/Design-System%2C-Desktop?type=design&node-id=8586-1421&mode=design&t=6RFLyrABcjt0f5u0-4)

Aussi, le bouton ADD TABS semble pas mal différent du bouton tertiaire que l'on utilise habituellement. Serait-ce possible de réutiliser le bouton tertiaire afin que l'on ait l'effet hover, la forme ronde et tout les autres particularités que nos boutons ont?

C'est peut-être hors scope mais je vais faire le commentaire quand même. Dans le tablist, on a un padding-left de 32px mais celui-ci devrait être de 16px afin que le contenu de la section soit aligné avec le label du premier Tab. Voir sur l'exemple suivant le décalage:
![image](https://github.com/kronostechnologies/design-elements/assets/52255752/a789bc9a-8611-4043-b9da-3b9035ef6f97)

Finalement, lorsqu'un Tab est sélectionné, il est toujours encerclé de son état focus. Est-ce que c'est vraiment nécessaire?

Je pense que ça fait pas mal le tour de mon côté

savutsang commented:
> Finalement, lorsqu'un Tab est sélectionné, il est toujours encerclé de son état focus. Est-ce que c'est vraiment nécessaire?

Fixed

meriouma commented on packages/react/src/components/tabs/tab-button.tsx:
```diff
@@ -95,9 +105,8 @@ interface TabButtonProps {
-
     onClick(): void;
-
+    onDelete?(tabId: string): void;
```
`onRemove`?

meriouma commented on packages/react/src/components/tabs/tabs.tsx:
```diff
@@ -139,11 +154,18 @@ export const Tabs: VoidFunctionComponent<Props> = ({
+                {onAddTab && (
+                    <Button buttonType='tertiary' type="button" onClick={onAddTab}>
+                        <AddIcon name="plusSign" size='16' aria-hidden="true" focusable={false} />
+                        {t('addTabs')}
```
Le bouton sert à ajouter un seul tab
```suggestion
                        {t('addTab')}
```

meriouma commented on packages/react/src/i18n/translations.ts:
```diff
@@ -106,6 +106,10 @@ export const Translations = {
         },
+        tabs: {
+            addTabs: 'Add Tabs',
+            dismissTab: 'Dismiss {{label}} tab',
```
On a potentiellement 2 nomenclatures pour le même besoin (delete/dismiss). Je mettrais plutôt `remove` partout je pense.

savutsang replied:
Je vais mettre "Remove",  dismiss semble un peu. Dismiss semble mieux adapte dans le cas t'a de quoi de nouveau et tu veux l'ignorer, comme un notification.
@LarryMatte

meriouma commented on packages/react/src/components/tabs/tab-panel.tsx:
```diff
@@ -8,15 +8,16 @@ interface TabPanelProps {
     contained?: boolean;
     hidden: boolean,
     id: string,
+    isGlobal?: boolean;
```
```suggestion
    global?: boolean;
```

meriouma reviewed (commented):
Petit review, je vais refaire une passe après les validations de Marc et/ou Larry.

savutsang commented:
> C'est peut-être hors scope mais je vais faire le commentaire quand même. Dans le tablist, on a un padding-left de 32px mais celui-ci devrait être de 16px afin que le contenu de la section soit aligné avec le label du premier Tab.

Fixed

maboilard reviewed (dismissed):
C'est comme si j'étais dans Figma! Excellent boulot Savut!! :)

savutsang commented on packages/react/src/components/tabs/tab-panel.tsx:
```diff
@@ -11,12 +11,10 @@ interface TabPanelProps {
 
 const StyledDiv = styled.div<{ $contained?: boolean }>`
-    border: ${({ $contained }) => ($contained ? '1px solid #878F9A' : 'none')}; /* TODO change with next thematization */
+    border: ${({ $contained }) => ($contained ? '1px solid #DBDEE1' : 'none')}; /* TODO change with next thematization */
```
D'apres ce que je comprend, le "Normal" dans Storybook c'est plus pour dire default ou basic exemple, c'est le meme nom qui est utilise dans tous les stories. Ca ne fait pas reference au Tabs de type normal.

Le background blanc des panel a ete ajoute.

savutsang commented:
> * J'enlèverais la possibilité de focus (touche Tab du clavier) sur les "x" des tabs qui ne sont pas actives.

Fixed

> * Dans la story "normal" quand j'utilise la touche Tab du clavier, le focus passe du **Tab** au **TabPanel** et après il devrait tomber sur le bouton de sorting dans le tableau mais il ne s'y rend pas. Comme si la navigation avec la touche Tab skippait ce qui se trouve dans le **TabPanel**.

Ca rentre dedans, le activeElement est le bouton du sorting, mais il n'y a pas de style focus sur ce bouton. Ca serait un bug du Table.  Si tu check dans le tab Note, il rentre dans le Textarea.

savutsang commented on packages/react/src/components/tabs/tabs.tsx:
```diff
@@ -7,18 +7,94 @@ import {
-    padding: ${({ $isGlobal }) => ($isGlobal ? '0' : '0 0 0 var(--spacing-4x)')};
+    gap: var(--spacing-half);
+    height: var(--size-2halfx);
+    overflow-x: auto;
```
Il me reste un bug, avec le overflow pour le scrollable, les style du focus sont coupe.

![image](https://github.com/kronostechnologies/design-elements/assets/3039571/70896782-c115-4a31-8b1b-8a84f45f2836)

LarryMatte replied:
Je ne crois pas qu'il y a grand chose que l'on puisse faire ici. J'ai fouillé pas mal et je n'ai vraiment rien trouvé. C'est pas mal le comportement by default dans chrome.
Ce que je propose serait d'y aller avec un focus strictement à l'intérieur (inset) un peu comme ça: https://codepen.io/larrymatte/pen/abxWBXz

LarryMatte commented on packages/react/src/components/tabs/tabs.tsx:
```diff
@@ -7,18 +7,93 @@ import {
+    white-space: nowrap;
+`;
+
+const ScrollButton = styled(Button) <{ $global?: boolean; $position: 'left' | 'right' }>`
```
Pour le problème avec les tabs qui est en partie visible derrière les bouton `ScrollButton` quand on scroll, une option serait de ne pas faire passer le scroll derrière, c'est à dire, ne pas positionner les boutons `ScrollButton` en `position: absolute;`.
Si tu ajoutes un `display: flex;` au `TabButtonsContainer` et enlève les `position: absolute;` au `ScrollButton`, je crois que ça vient pas mal régler le problème.

savutsang replied:
On peut l'ignorer, c'est un probleme qui arrive seulement lors de ma tentative de faire fonctionner le focus. Si on fait le inset focus a place, ca ne se reproduira pas. C'est necessaire de mettre les buttons en absolute ici parce que les tabs commence a la meme place que les buttons. Et lorsque les buttons disparaissent, on ne veut pas que ca fasse un saut.

savutsang commented:
J'ai fait le focus inside. J'ai pas rebase sur la branche de Henri car ce n'est que 2 lignes et puisque son PR vas aussi prendre un temps inconnu, je veux pas que les reviews soit attache.

meriouma commented on packages/react/src/components/tabs/tab-button.tsx:
```diff
@@ -1,90 +1,100 @@
-            position: absolute;
-            width: 100%;
+            &::after {
+                background-color: ${theme.component['tabs-tab-active-indicator-color']} !important;
```
Pourquoi le `!important` ici?

savutsang replied:
C'est que le active ici se fait override par le hover qui est sur une couche plus haut avec une specificite plus grande. Cela est necessaire pour que le style hover s'applique aussi lorsque on hover sur le X. J'aime pas trop ca non plus mais ca me semble le moins pire des options.

```
<div TabItem> // hover
    <button TabButton> // active, focus
    <button remove>
</div>
```

meriouma commented on packages/react/src/components/tabs/tab-panel.tsx:
```diff
@@ -5,35 +5,42 @@ import { focus, focusVisibleReset } from '../../utils/css-state';
-    contained,
     hidden,
     id,
+    global,
```
Valeur par défaut? Aussi tu peux le remonter avant `hidden` pour garder l'ordre alpha
```suggestion
    global = false,
```

meriouma commented on packages/react/src/components/tabs/tabs.tsx:
```diff
@@ -27,6 +102,7 @@ export interface Tab {
     leftIcon?: IconName;
     rightIcon?: IconName;
     panelContent: ReactNode;
+    onRemove?(tabId: string): void;
```
Est-ce que le `onRemove` devrait être sur les props pour Tabs plutôt?
Sinon ici comme tu es dans un tab spécifique, le `tabId` me semble superflu vu que tu as l'id dans l'objet et que le `onRemove` est spécifique à un tab.
Je pense que ça aurait plus de sens de le remonter sur le component.

meriouma commented on packages/react/src/hooks/use-scrollable.ts:
```diff
@@ -0,0 +1,68 @@
+interface UseScrollableOptions {
+    scrollableElement: RefObject<HTMLElement>;
+    scrollByPercent: number;
+    onScroll: (params: { atStartX: boolean; atEndX: boolean; }) => void;
```
Peut-être extraire une interface pour `params`

meriouma commented on packages/react/src/hooks/use-scrollable.ts:
```diff
@@ -0,0 +1,68 @@
+}
+
+interface UseScrollableReturns {
+    scrollToLeft: () => void;
+    scrollToRight: () => void;
```
C'est une préférence personnelle
```suggestion
    scrollToLeft(): void;
    scrollToRight(): void;
```

meriouma commented on packages/react/src/hooks/use-scrollable.ts:
```diff
@@ -0,0 +1,68 @@
+        scrollableElement.current.scrollTo({ left: newPosX, behavior: 'smooth' });
+    };
+
+    return {
```
Je suis toujours un peu incertain, mais je me demande si on doit pas memoized le retour d'un custom hook quand c'est un objet.

savutsang commented:
Tout a ete fixe, pret pour un review.

savutsang commented:
Tout a ete fixe, pret pour un review.

savutsang commented:
Tout a ete fixe, pret pour un review.

LarryMatte commented on packages/react/src/components/tabs/tab-button.tsx:
```diff
@@ -1,91 +1,100 @@
+    right: var(--spacing-1halfx);
+    top: 50%;
+    transform: translateY(-50%);
+    width: var(--size-1x);
```
Le bouton est un peu squeezed. Si tu remplaces `var(--size-1x)` par `auto`, ca devrait etre correct.
<img width="119" alt="Screenshot 2024-04-24 at 2 18 56 PM" src="https://github.com/kronostechnologies/design-elements/assets/3722754/d1999f81-be49-4111-870a-2b1976db9b90">

savutsang replied:
C'est bete... j'ai decouvert que IconButton a un prop `size="small"`, j'ai remplace tout mon CSS par ca a place.

LarryMatte reviewed (changes requested):
Quelques points:

1. Quand le focus est sur le bouton “delete” et qu’on supprime la Tab ou qu’on supprime la Tab active avec la souris, le focus disparait et il n’y a plus de Tab active laissant ainsi un vide sous les tabs (il n'y a pas de tabpanel d'affiché i.e., aucun contenu sous les tabs) ET il n’est plus possible de focus sur les Tabs qui sont inactive. Il faudrait que le focus bouge vers la Tab suivante, la rendant active au même moment.
2. Quand le focus est sur le bouton “delete” de la tab active, il faudrait pouvoir bouger à la prochaine Tab en utilisant la flèche du clavier (comme quand le focus est sur la tab elle même).
3. Le `aria-label` du bouton pour supprimer une Tab n’a pas le label de la tab dedans, il y a seulement `dismissTab`. Il faudrait que ce soit `Dismiss [nom du tab] Tab`. Donc `dismiss first button tab`.

hebernardEquisoft commented on packages/react/src/themes/tokens/component/tab-tokens.ts:
```diff
@@ -22,16 +23,17 @@ export type TabTokenMap = {
-    'tabs-tab-text-color': 'color-neutral-50',
-    'tabs-tab-selected-background-color': 'color-white',
-    'tabs-tab-selected-border-color': 'color-neutral-50',
+    'tabs-background-color': 'color-neutral-05',
```
```suggestion
    'tab-background-color': 'color-neutral-05',
    'tab-global-background-color': 'color-white',
    'tab-border-color': 'color-neutral-15',
    'tab-icon-color': 'color-neutral-65',
    'tab-text-color': 'color-neutral-65',
    
    'tab-indicator-hover-background-color': 'color-neutral-15',
    'tab-hover-text-color': 'color-neutral-65',
    
    'tab-indicator-pressed-background-color': 'color-brand-70',
    'tab-pressed-text-color': 'color-neutral-90',
    
    'tab-indicator-selected-background-color': 'color-brand-50',
    'tab-selected-text-color': 'color-neutral-90',

    'tab-panel-border-color': 'color-neutral-05',
    'tab-panel-background-color': 'color-white',
```

faudrait demander à @maboilard si active s'inscrit à leur language de design ou si ça serait plus pressed...

maboilard replied:
Actuellement, c'est `Active` mais j'ai l'impression que `Pressed` pourrait être confondant si à un certain point, on décide d'appliquer un nouvel état Pressed au clic. Est-ce que `Selected` serait plus approprié?

hebernardEquisoft replied:
Je crois que dans ce cas ça serait `active` puisque `selected` est l'état après avoir pressed/active

LarryMatte replied:
Si on essaie d'utiliser les termes déjà dispo en web.

`active` en css = quand tu cliques sur l'élément. i.e., tu peux changer le style d'un élément clickable au moment tu le trigger avec la souris  https://developer.mozilla.org/en-US/docs/Web/CSS/:active

`aria-selected` en a11y = This attribute is used to indicate which elements within single-selection and multiple-selection composite widgets are selected. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-selected

`aria-pressed` en a11y = This attribute indicates the current "pressed" state of a toggle button. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed

Je vous laisse juger quel terme utiliser.

savutsang replied:
Selon les docs, "selected" c'est quand tu l'a choisi (le current Tabs).  "pressed" c'est presque la meme chose mais pour les toggle buttons.

J'ai mis "active" pour quand tu presse dessus a cause du nom en CSS.  Cependant j'ai trouve plusieurs DS qui utilise "active" comme l'equivalent de "selected".

active: Bootstrap, Fluent, Semantic-UI, Ants
selected: Material-UI, Carbon, Atlassian, Adobe

La majorite n'ont pas de style pour quand tu presse dessus, ils ont juste un hover, ou alors c'est fait en CSS, alors j'ai pas trouve comment ils nomme ca.

Si on va avec les specs CSS et A11y, moi j'irais pour:

current tab: selected
mouse pressed: active

savutsang commented:
> * Quand le focus est sur le bouton “delete” et qu’on supprime la Tab ou qu’on supprime la Tab active avec la souris, le focus disparait et il n’y a plus de Tab active laissant ainsi un vide sous les tabs (il n'y a pas de tabpanel d'affiché i.e., aucun contenu sous les tabs) ET il n’est plus possible de focus sur les Tabs qui sont inactive. Il faudrait que le focus bouge vers la Tab suivante, la rendant active au même moment.

Fixed

> * Quand le focus est sur le bouton “delete” de la tab active, il faudrait pouvoir bouger à la prochaine Tab en utilisant la flèche du clavier (comme quand le focus est sur la tab elle même).

Fixed

> * Le `aria-label` du bouton pour supprimer une Tab n’a pas le label de la tab dedans, il y a seulement `dismissTab`. Il faudrait que ce soit `Dismiss [nom du tab] Tab`. Donc `dismiss first button tab`.

Fixed

savutsang commented:
Tout a ete fixe, pret pour un autre autre review.

LarryMatte commented:
> * Il faudrait que le focus bouge vers la Tab suivante, la rendant active au même moment.

@savutsang 
Il reste à ajouter l'état focus sur le Tab active après la suppression d'une Tab.

savutsang commented:
Tout a ete fixe, pret pour un autre autre autre review.

savutsang commented:
> @savutsang Il reste à ajouter l'état focus _i.e._, le contour qui est highlighted, sur le Tab active après la suppression d'une Tab.

Tout a ete fixe, pret pour un autre autre autre review.

savutsang commented:
Tout a ete fixe, pret pour un autre autre autre review.

LarryMatte reviewed (approved):
LGTM
